File: JasSwift/Sources/Geometry/HitTest.swift
import Foundation

/// Geometry helpers for precise hit-testing.
///
/// Pure-geometry functions used by the controller for marquee selection,
/// element intersection tests, and control-point queries.  These do not
/// depend on the document model — only on element geometry.

// MARK: - Primitive geometry

public func pointInRect(_ px: Double, _ py: Double,
                        _ rx: Double, _ ry: Double, _ rw: Double, _ rh: Double) -> Bool {
    rx <= px && px <= rx + rw && ry <= py && py <= ry + rh
}

private func cross(_ ox: Double, _ oy: Double, _ ax: Double, _ ay: Double,
                   _ bx: Double, _ by: Double) -> Double {
    (ax - ox) * (by - oy) - (ay - oy) * (bx - ox)
}

private func onSegment(_ px1: Double, _ py1: Double, _ px2: Double, _ py2: Double,
                       _ qx: Double, _ qy: Double) -> Bool {
    min(px1, px2) <= qx && qx <= max(px1, px2) &&
    min(py1, py2) <= qy && qy <= max(py1, py2)
}

public func segmentsIntersect(_ ax1: Double, _ ay1: Double, _ ax2: Double, _ ay2: Double,
                              _ bx1: Double, _ by1: Double, _ bx2: Double, _ by2: Double) -> Bool {
    let d1 = cross(bx1, by1, bx2, by2, ax1, ay1)
    let d2 = cross(bx1, by1, bx2, by2, ax2, ay2)
    let d3 = cross(ax1, ay1, ax2, ay2, bx1, by1)
    let d4 = cross(ax1, ay1, ax2, ay2, bx2, by2)
    if ((d1 > 0 && d2 < 0) || (d1 < 0 && d2 > 0)) &&
       ((d3 > 0 && d4 < 0) || (d3 < 0 && d4 > 0)) { return true }
    let eps = 1e-10
    if abs(d1) < eps && onSegment(bx1, by1, bx2, by2, ax1, ay1) { return true }
    if abs(d2) < eps && onSegment(bx1, by1, bx2, by2, ax2, ay2) { return true }
    if abs(d3) < eps && onSegment(ax1, ay1, ax2, ay2, bx1, by1) { return true }
    if abs(d4) < eps && onSegment(ax1, ay1, ax2, ay2, bx2, by2) { return true }
    return false
}

public func segmentIntersectsRect(_ x1: Double, _ y1: Double, _ x2: Double, _ y2: Double,
                                  _ rx: Double, _ ry: Double, _ rw: Double, _ rh: Double) -> Bool {
    if pointInRect(x1, y1, rx, ry, rw, rh) { return true }
    if pointInRect(x2, y2, rx, ry, rw, rh) { return true }
    let edges: [(Double, Double, Double, Double)] = [
        (rx, ry, rx + rw, ry),
        (rx + rw, ry, rx + rw, ry + rh),
        (rx + rw, ry + rh, rx, ry + rh),
        (rx, ry + rh, rx, ry),
    ]
    return edges.contains { e in
        segmentsIntersect(x1, y1, x2, y2, e.0, e.1, e.2, e.3)
    }
}

public func rectsIntersect(_ ax: Double, _ ay: Double, _ aw: Double, _ ah: Double,
                           _ bx: Double, _ by: Double, _ bw: Double, _ bh: Double) -> Bool {
    ax < bx + bw && ax + aw > bx && ay < by + bh && ay + ah > by
}

public func circleIntersectsRect(_ cx: Double, _ cy: Double, _ r: Double,
                                 _ rx: Double, _ ry: Double, _ rw: Double, _ rh: Double,
                                 filled: Bool) -> Bool {
    let closestX = max(rx, min(cx, rx + rw))
    let closestY = max(ry, min(cy, ry + rh))
    let distSq = pow(cx - closestX, 2) + pow(cy - closestY, 2)
    if !filled {
        let corners = [(rx, ry), (rx + rw, ry), (rx + rw, ry + rh), (rx, ry + rh)]
        let maxDistSq = corners.map { pow(cx - $0.0, 2) + pow(cy - $0.1, 2) }.max()!
        return distSq <= r * r && r * r <= maxDistSq
    }
    return distSq <= r * r
}

public func ellipseIntersectsRect(_ cx: Double, _ cy: Double, _ erx: Double, _ ery: Double,
                                  _ rx: Double, _ ry: Double, _ rw: Double, _ rh: Double,
                                  filled: Bool) -> Bool {
    if erx == 0 || ery == 0 { return false }
    return circleIntersectsRect(cx / erx, cy / ery, 1.0,
                                rx / erx, ry / ery, rw / erx, rh / ery,
                                filled: filled)
}

// MARK: - Element-level queries

public func segmentsOfElement(_ elem: Element) -> [(Double, Double, Double, Double)] {
    switch elem {
    case .line(let v):
        return [(v.x1, v.y1, v.x2, v.y2)]
    case .rect(let v):
        let x = v.x, y = v.y, w = v.width, h = v.height
        return [(x, y, x+w, y), (x+w, y, x+w, y+h),
                (x+w, y+h, x, y+h), (x, y+h, x, y)]
    case .polyline(let v):
        guard v.points.count >= 2 else { return [] }
        return (0..<v.points.count-1).map { i in
            (v.points[i].0, v.points[i].1, v.points[i+1].0, v.points[i+1].1)
        }
    case .polygon(let v):
        guard v.points.count >= 2 else { return [] }
        var segs = (0..<v.points.count-1).map { i in
            (v.points[i].0, v.points[i].1, v.points[i+1].0, v.points[i+1].1)
        }
        let last = v.points.last!, first = v.points.first!
        segs.append((last.0, last.1, first.0, first.1))
        return segs
    case .path(let v):
        let pts = flattenPathCommands(v.d)
        guard pts.count >= 2 else { return [] }
        return (0..<pts.count-1).map { i in
            (pts[i].0, pts[i].1, pts[i+1].0, pts[i+1].1)
        }
    default:
        return []
    }
}

public func allCPs(_ elem: Element) -> Set<Int> {
    Set(0..<elem.controlPointCount)
}

// TODO: This ignores the element's transform. If an element has a non-identity
// transform, its visual position differs from its raw coordinates. To fix,
// inverse-transform the selection rect into the element's local coordinate
// space before testing (inheriting transforms from parent groups).
public func elementIntersectsRect(_ elem: Element,
                                  _ rx: Double, _ ry: Double, _ rw: Double, _ rh: Double) -> Bool {
    switch elem {
    case .line(let v):
        return segmentIntersectsRect(v.x1, v.y1, v.x2, v.y2, rx, ry, rw, rh)
    case .rect(let v):
        if v.fill != nil {
            return rectsIntersect(v.x, v.y, v.width, v.height, rx, ry, rw, rh)
        }
        return segmentsOfElement(elem).contains { s in
            segmentIntersectsRect(s.0, s.1, s.2, s.3, rx, ry, rw, rh)
        }
    case .circle(let v):
        return circleIntersectsRect(v.cx, v.cy, v.r, rx, ry, rw, rh, filled: v.fill != nil)
    case .ellipse(let v):
        return ellipseIntersectsRect(v.cx, v.cy, v.rx, v.ry, rx, ry, rw, rh, filled: v.fill != nil)
    case .polyline(let v):
        if v.fill != nil {
            let b = elem.bounds
            return rectsIntersect(b.x, b.y, b.width, b.height, rx, ry, rw, rh)
        }
        return segmentsOfElement(elem).contains { s in
            segmentIntersectsRect(s.0, s.1, s.2, s.3, rx, ry, rw, rh)
        }
    case .polygon(let v):
        if v.fill != nil {
            if v.points.contains(where: { pointInRect($0.0, $0.1, rx, ry, rw, rh) }) {
                return true
            }
            return segmentsOfElement(elem).contains { s in
                segmentIntersectsRect(s.0, s.1, s.2, s.3, rx, ry, rw, rh)
            }
        }
        return segmentsOfElement(elem).contains { s in
            segmentIntersectsRect(s.0, s.1, s.2, s.3, rx, ry, rw, rh)
        }
    case .path(let v):
        let segs = segmentsOfElement(elem)
        if v.fill != nil {
            let endpoints = segs.flatMap { [(s: $0.0, t: $0.1), (s: $0.2, t: $0.3)] }
            if endpoints.contains(where: { pointInRect($0.s, $0.t, rx, ry, rw, rh) }) {
                return true
            }
            return segs.contains { s in
                segmentIntersectsRect(s.0, s.1, s.2, s.3, rx, ry, rw, rh)
            }
        }
        return segs.contains { s in
            segmentIntersectsRect(s.0, s.1, s.2, s.3, rx, ry, rw, rh)
        }
    case .text:
        let b = elem.bounds
        return rectsIntersect(b.x, b.y, b.width, b.height, rx, ry, rw, rh)
    default:
        let b = elem.bounds
        return rectsIntersect(b.x, b.y, b.width, b.height, rx, ry, rw, rh)
    }
}

File: jas/geometry/hit_test.py
"""Geometry helpers for precise hit-testing.

Pure-geometry functions used by the controller for marquee selection,
element intersection tests, and control-point queries.  These do not
depend on the document model — only on element geometry.
"""

from __future__ import annotations

from geometry.element import (
    Circle, Element, Ellipse, Group, Layer, Line, Path, Polygon, Polyline,
    Rect, Text,
    control_point_count,
    flatten_path_commands,
)


# ---------------------------------------------------------------------------
# Primitive geometry
# ---------------------------------------------------------------------------

def point_in_rect(px: float, py: float,
                  rx: float, ry: float, rw: float, rh: float) -> bool:
    return rx <= px <= rx + rw and ry <= py <= ry + rh


def _cross(ox: float, oy: float, ax: float, ay: float,
           bx: float, by: float) -> float:
    return (ax - ox) * (by - oy) - (ay - oy) * (bx - ox)


def _on_segment(px1: float, py1: float, px2: float, py2: float,
                qx: float, qy: float) -> bool:
    return (min(px1, px2) <= qx <= max(px1, px2) and
            min(py1, py2) <= qy <= max(py1, py2))


def segments_intersect(ax1: float, ay1: float, ax2: float, ay2: float,
                       bx1: float, by1: float, bx2: float, by2: float) -> bool:
    d1 = _cross(bx1, by1, bx2, by2, ax1, ay1)
    d2 = _cross(bx1, by1, bx2, by2, ax2, ay2)
    d3 = _cross(ax1, ay1, ax2, ay2, bx1, by1)
    d4 = _cross(ax1, ay1, ax2, ay2, bx2, by2)
    if ((d1 > 0 and d2 < 0) or (d1 < 0 and d2 > 0)) and \
       ((d3 > 0 and d4 < 0) or (d3 < 0 and d4 > 0)):
        return True
    eps = 1e-10
    if abs(d1) < eps and _on_segment(bx1, by1, bx2, by2, ax1, ay1): return True
    if abs(d2) < eps and _on_segment(bx1, by1, bx2, by2, ax2, ay2): return True
    if abs(d3) < eps and _on_segment(ax1, ay1, ax2, ay2, bx1, by1): return True
    if abs(d4) < eps and _on_segment(ax1, ay1, ax2, ay2, bx2, by2): return True
    return False


def segment_intersects_rect(x1: float, y1: float, x2: float, y2: float,
                            rx: float, ry: float, rw: float, rh: float) -> bool:
    if point_in_rect(x1, y1, rx, ry, rw, rh):
        return True
    if point_in_rect(x2, y2, rx, ry, rw, rh):
        return True
    edges = [
        (rx, ry, rx + rw, ry),
        (rx + rw, ry, rx + rw, ry + rh),
        (rx + rw, ry + rh, rx, ry + rh),
        (rx, ry + rh, rx, ry),
    ]
    return any(segments_intersect(x1, y1, x2, y2, *e) for e in edges)


def rects_intersect(ax: float, ay: float, aw: float, ah: float,
                    bx: float, by: float, bw: float, bh: float) -> bool:
    return ax < bx + bw and ax + aw > bx and ay < by + bh and ay + ah > by


def circle_intersects_rect(cx: float, cy: float, r: float,
                           rx: float, ry: float, rw: float, rh: float,
                           filled: bool) -> bool:
    closest_x = max(rx, min(cx, rx + rw))
    closest_y = max(ry, min(cy, ry + rh))
    dist_sq = (cx - closest_x) ** 2 + (cy - closest_y) ** 2
    if not filled:
        corners = [(rx, ry), (rx + rw, ry), (rx + rw, ry + rh), (rx, ry + rh)]
        max_dist_sq = max((cx - cx2) ** 2 + (cy - cy2) ** 2 for cx2, cy2 in corners)
        return dist_sq <= r * r <= max_dist_sq
    return dist_sq <= r * r


def ellipse_intersects_rect(cx: float, cy: float, erx: float, ery: float,
                            rx: float, ry: float, rw: float, rh: float,
                            filled: bool) -> bool:
    if erx == 0 or ery == 0:
        return False
    return circle_intersects_rect(
        cx / erx, cy / ery, 1.0,
        rx / erx, ry / ery, rw / erx, rh / ery,
        filled,
    )


# ---------------------------------------------------------------------------
# Element-level queries
# ---------------------------------------------------------------------------

def segments_of_element(elem: Element) -> list[tuple[float, float, float, float]]:
    """Return the line segments that make up the visible drawn edges of an element."""
    match elem:
        case Line(x1=x1, y1=y1, x2=x2, y2=y2):
            return [(x1, y1, x2, y2)]
        case Rect(x=x, y=y, width=w, height=h):
            return [(x, y, x+w, y), (x+w, y, x+w, y+h),
                    (x+w, y+h, x, y+h), (x, y+h, x, y)]
        case Polyline(points=pts):
            return [(pts[i][0], pts[i][1], pts[i+1][0], pts[i+1][1])
                    for i in range(len(pts) - 1)] if len(pts) >= 2 else []
        case Polygon(points=pts):
            if len(pts) < 2:
                return []
            segs = [(pts[i][0], pts[i][1], pts[i+1][0], pts[i+1][1])
                    for i in range(len(pts) - 1)]
            segs.append((pts[-1][0], pts[-1][1], pts[0][0], pts[0][1]))
            return segs
        case Path(d=cmds):
            pts = flatten_path_commands(cmds)
            return [(pts[i][0], pts[i][1], pts[i+1][0], pts[i+1][1])
                    for i in range(len(pts) - 1)] if len(pts) >= 2 else []
        case _:
            return []


def all_cps(elem: Element) -> frozenset[int]:
    """Return a frozenset of all control point indices for an element."""
    return frozenset(range(control_point_count(elem)))


def element_intersects_rect(elem: Element,
                            rx: float, ry: float, rw: float, rh: float) -> bool:
    """Test whether the visible drawn portion of elem intersects the selection rect.

    TODO: This ignores the element's transform. If an element has a non-identity
    transform, its visual position differs from its raw coordinates. To fix,
    inverse-transform the selection rect into the element's local coordinate
    space before testing (inheriting transforms from parent groups).
    """
    match elem:
        case Line():
            return segment_intersects_rect(elem.x1, elem.y1, elem.x2, elem.y2,
                                            rx, ry, rw, rh)
        case Rect():
            if elem.fill is not None:
                return rects_intersect(elem.x, elem.y, elem.width, elem.height,
                                        rx, ry, rw, rh)
            return any(segment_intersects_rect(*seg, rx, ry, rw, rh)
                       for seg in segments_of_element(elem))

        case Circle():
            return circle_intersects_rect(elem.cx, elem.cy, elem.r,
                                           rx, ry, rw, rh,
                                           elem.fill is not None)
        case Ellipse():
            return ellipse_intersects_rect(elem.cx, elem.cy, elem.rx, elem.ry,
                                            rx, ry, rw, rh,
                                            elem.fill is not None)
        case Polyline():
            if elem.fill is not None:
                return rects_intersect(*elem.bounds(), rx, ry, rw, rh)
            return any(segment_intersects_rect(*seg, rx, ry, rw, rh)
                       for seg in segments_of_element(elem))

        case Polygon():
            if elem.fill is not None:
                pts = elem.points
                if any(point_in_rect(px, py, rx, ry, rw, rh) for px, py in pts):
                    return True
                return any(segment_intersects_rect(*seg, rx, ry, rw, rh)
                           for seg in segments_of_element(elem))
            return any(segment_intersects_rect(*seg, rx, ry, rw, rh)
                       for seg in segments_of_element(elem))

        case Path():
            if elem.fill is not None:
                segs = segments_of_element(elem)
                endpoints = [(s[0], s[1]) for s in segs] + [(s[2], s[3]) for s in segs]
                if any(point_in_rect(px, py, rx, ry, rw, rh) for px, py in endpoints):
                    return True
                return any(segment_intersects_rect(*seg, rx, ry, rw, rh) for seg in segs)
            return any(segment_intersects_rect(*seg, rx, ry, rw, rh)
                       for seg in segments_of_element(elem))

        case Text():
            return rects_intersect(*elem.bounds(), rx, ry, rw, rh)

        case _:
            return rects_intersect(*elem.bounds(), rx, ry, rw, rh)
